File: TextExercise/12-1/Java/Item.java
/**
 * 貪欲アルゴリズムの品物に関するJava実装
 */
public class Item implements HasWeightAndValue{
    private String name;
    private int value;
    private float weight;

    /**
     * コンストラクタ。Immutable Objectにするのでここでしか値はセットされない。
     * @param name 名称
     * @param value 価値
     * @param weight 重さ
     */
    public Item(String name, int value, int weight) {
        this.name = name;
        this.value = value;
        this.weight = weight;
    }

    /**
     * 名称を取得する。
     * @return 名称
     */
    public String getName() {
        return this.name;
    }

    /**
     * 価値を取得する。
     * @return 価値
     */
    @Override
    public float getValue() {
        return (float)this.value;
    }

    /**
     * 質量を取得する。
     * @return 質量
     */
    @Override
    public float getWeight() {
        return (float)this.weight;
    }    
}

File: TextExercise/12-1/Python/item.py
"""貪欲アルゴリズムで使用するItemクラスとそれにまつわる関数の定義"""

class Item():
    """ナップサックに詰める対象の品物"""
    def __init__(self, n, v, w):
        """初期化。weightについては割り算の分母になるのと質量なので０以下の値は入れない"""
        self.__name = n
        self.__value = v
        if float(w) <= 0.0:
            raise ValueError(v)
        self.__weight = w

    def getName(self):
        """名前取得"""
        return self.__name

    def getWeight(self):
        """質量取得"""
        return self.__weight

    def getValue(self):
        """価値取得"""
        return self.__value

    def __str__(self):
        """文字列表現"""
        return f'[{self.name}] weight = {self.weight}, value = {self.value}'

def value(item):
    """「価値の高いもの」を基準に取る場合"""
    return item.getValue()

def weightInverse(item):
    """「重量の軽いもの」を基準に取る場合"""
    return 1.0/item.getWeight()

def density(item):
    """「価値/重量比」を基準に取る場合"""
    return item.getValue()/item.getWeight()
